handler/oauth2/flow_generic_code_token.go

```diff
@@ -0,0 +1,244 @@
+// Copyright © 2024 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package oauth2
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/ory/fosite/storage"
+
+	"github.com/ory/x/errorsx"
+
+	"github.com/ory/fosite"
+)
+
+// CodeTokenEndpointHandler handles the differences between Authorize code grant and extended grant types.
+type CodeTokenEndpointHandler interface {
+	// ValidateGrantTypes validates the authorization grant type.
+	ValidateGrantTypes(ctx context.Context, requester fosite.AccessRequester) error
+
+	// ValidateCode validates the code used in the authorization flow.
+	ValidateCode(ctx context.Context, requester fosite.AccessRequester, code string) error
+
+	// GetCodeAndSession retrieves the code, the code signature, and the request session.
+	GetCodeAndSession(ctx context.Context, requester fosite.AccessRequester) (code string, signature string, authorizeRequest fosite.Requester, err error)
```

Comment:
I think we should avoid using the term `code` as different kinds of tokens can be used on the token endpoints (auth_code, device_code, refresh_token)

```suggestion
	// ValidateToken validates the token that is used to retrieve the access token.
	ValidateToken(ctx context.Context, requester fosite.AccessRequester, code string) error

	// GetTokenAndSession retrieves the token, the token signature, and the request session.
	GetTokenAndSession(ctx context.Context, requester fosite.AccessRequester) (code string, signature string, authorizeRequest fosite.Requester, err error)
```